rclrs_examples/src/rclrs_publisher.rs

```diff
@@ -1,22 +1,26 @@
-extern crate rclrs;
-extern crate std_msgs;
+use rclrs;
+use std_msgs;
 
-fn main() {
-    rclrs::init().unwrap();
+fn main() -> rclrs::RclResult {
+    let context = rclrs::Context::default();
+
+    let node = context.create_node("minimal_publisher")?;
 
-    let node = rclrs::create_node("minimal_publisher");
     let publisher =
-        node.create_publisher::<std_msgs::msg::String>("topic", rclrs::qos::QOS_PROFILE_DEFAULT);
+        node.create_publisher::<std_msgs::msg::String>("topic", rclrs::QOS_PROFILE_DEFAULT)?;
 
-    let mut message: std_msgs::msg::String = Default::default();
+    let mut message = std_msgs::msg::String::default();
 
     let mut publish_count: u32 = 1;
 
-    while rclrs::ok() {
+    while context.ok() {
```

Comment:
I'd keep rclrs:::ok for symmetry with the other client libraries.

Reply:
From [https://index.ros.org//doc/ros2/Releases/Release-Crystal-Clemmys/](https://index.ros.org//doc/ros2/Releases/Release-Crystal-Clemmys/).

> However, you may just continue to offer a single, global init and shutdown in your client library, and just store a single global context object.

As it is optional I was thinking about adding the global context object later.

Reply:
It's fine if the context is not global for the time being, we can address that in the future. What I meant is that all the client libraries have an 'init` function that optionally take a context (or use the global one implicitly) and all functions are attached to the root namespace (e.g. rclcpp::init, rclpy.init, etc.)

Moreover, the concept of the context as defined in ROS2 differs slightly from what's presented in this pull request. It's not the context that calls `rcl_init`, but the opposite:

https://github.com/ros2/rcl/blob/master/rcl/src/rcl/init.c#L59

so in order to have mimic the other client libraries, we need to have a `rclrs::init` function that accepts a context and all functions are done at the rclrs level (e.g. spin,  ok, etc.)

Reply:
> Moreover, the concept of the context as defined in ROS2 differs slightly from what's presented in this pull request. It's not the context that calls rcl_init, but the opposite

It looks like rclcpp context calls rcl_init.

[https://github.com/ros2/rclcpp/blob/a54a32915313960497bda2d601cdf98e9f2e9bf0/rclcpp/src/rclcpp/context.cpp#L89](https://github.com/ros2/rclcpp/blob/a54a32915313960497bda2d601cdf98e9f2e9bf0/rclcpp/src/rclcpp/context.cpp#L89)

> in order to have mimic the other client libraries, we need to have a rclrs::init function that accepts a context and all functions are done at the rclrs level

rclpy.init does take an optional context but rclcpp::init does not.

[https://github.com/ros2/rclcpp/blob/e32d17980dcf761712cb854b3703336c380a570e/rclcpp/src/rclcpp/utilities.cpp#L28](https://github.com/ros2/rclcpp/blob/e32d17980dcf761712cb854b3703336c380a570e/rclcpp/src/rclcpp/utilities.cpp#L28)

How much of a requirement do you think it is to mimic another client interface ? I think it is an important question to settle since it can considerably restrain the design space in the future.

Reply:
> How much of a requirement do you think it is to mimic another client interface ? I think it is an important question to settle since it can considerably restrain the design space in the future.


I'd say it's pretty important that rclrs follows the other client libraries, if only to help users switch to Rust if they know other ROS2 libraries. There's only a handful of methods common to all the libraries (i.e. init, createX, destroyX), so apart from that, we have enough freedom to implement the API in any way we want.

Reply:
@lelongg did you have time to look into making the `rclrs` API similar to the other clients'? Thanks.

Reply:
I didn't make the change yet. What API do you suggest ? We can mimic the rclcpp API by making rclrs::ok take a context object but rclcpp::init does not and so we cannot add rclrs::init until we get the global context.